devito/ir/support/utils.py

```diff
@@ -86,16 +86,9 @@ def detect_flow_directions(exprs):
     mapper.update({d: {Any} for d in flatten(i.aindices for i in reads + writes)
                    if d is not None and d not in mapper})
 
-    # Add in stepping dimensions (just in case they haven't been detected yet)
-    # note: stepping dimensions may force a direction on the parent
-    assert all(v == {Any} or mapper.get(k.parent, v) in [v, {Any}]
-               for k, v in mapper.items() if k.is_Stepping)
+    # Add in derived-dimensions parents, in case they haven't been detected yet
```

Comment:
Couldn't put it above. Line 82:
`for d in dimensions[dimensions.index(d) + 1:]`
this fails as `d` is referenced before initialization.

Reply:
Fails where/when? It should pick the last d from the previus loop

Reply:
When `dimensions` is empty

Reply:
There is few other issue with `detect_flow_directions` I'm trying to create example for

Reply:
I'm actually not sure wht this two lines (82-83) do as all `d` in `dimension` is added to the mapper in the previous loop.

Reply:
> When dimensions is empty

good spot

Reply:
> I'm actually not sure wht this two lines (82-83) do as all d in dimension is added to the mapper in the previous loop.

No, in the previous loop you ``break`` ASAP